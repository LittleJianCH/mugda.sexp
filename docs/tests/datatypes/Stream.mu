(import "./Nat.mu" Nat zero)

;; `Stream` is an example of a coinductive type.
;; Coinductive types are not required to be well-founded,
;; i.e. their inhabitants do not need to have finite height.

;; For writing simple examples, we only define `Stream` of `Nat`.

;; There are no `Stream` objects of finite height.

(codata Stream Type
  [cons (-> Nat Stream Stream)])

(fn zeroes Stream
  [(zeroes) (cons zero zeroes)])

(fn head (-> Stream Nat)
  [((cons x xs)) x])
