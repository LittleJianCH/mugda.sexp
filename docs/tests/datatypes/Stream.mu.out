(((fn [(x (zero)) TODO] [(x (add1 y)) TODO]) zero) zero)
